internal/fields/mappings.go

```diff
@@ -0,0 +1,577 @@
+	// - If the same mapping exists in both, but they have different "type", there is some issue
+	//     - Could this happen?
+	//     - Should those documents being rejected directly in this case?
+	// - If there is a new mapping,
+	//     - It could come from a ECS definition, compare that mapping with the ECS field definitions
+	//     - Does this come from some dynamic template? ECS componente template or dynamic templates defined in the package? This mapping is valid
+	//         - conditions found in current dynamic templates: match, path_match, path_unmatch, match_mapping_type, unmatch_mapping_type
+	//     - if it does not match, there should be some issue and it should be reported
+	//     - If the mapping is a constant_keyword type (e.g. data_stream.dataset), how to check the value?
+	//         - if the constant_keyword is defined in the preview, it should be the same
+	if diff := cmp.Diff(actualMappings, previewMappings, transformJSON); diff == "" {
+		logger.Debugf("No changes found in mappings")
+		return errs.Unique()
+	}
+
+	var rawPreview mappingDefinitions
+	err = json.Unmarshal(previewMappings, &rawPreview)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("failed to unmarshal preview mappings (index template %s): %w", v.indexTemplateName, err))
+		return errs.Unique()
+	}
+	var rawActual mappingDefinitions
+	err = json.Unmarshal(actualMappings, &rawActual)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("failed to unmarshal actual mappings (data stream %s): %w", v.dataStreamName, err))
+		return errs.Unique()
+	}
+
+	mappingErrs := compareMappings("", rawPreview, rawActual, v.Schema, v.LocalSchema)
+	errs = append(errs, mappingErrs...)
+
+	if len(errs) > 0 {
+		return errs.Unique()
+	}
+
+	return nil
+}
+
+func mappingParameter(field string, definition mappingDefinitions) string {
+	fieldValue, ok := definition[field]
+	if !ok {
+		return ""
+	}
+	value, ok := fieldValue.(string)
+	if !ok {
+		return ""
+	}
+	return value
+}
+
+func isLocalFieldTypeArray(field string, schema []FieldDefinition) bool {
+	definition := findElementDefinitionForRoot("", field, schema)
+	if definition == nil {
+		return false
+	}
+	return definition.Type == "array"
+}
+
+func isEmptyObject(definition mappingDefinitions) bool {
+	// Example:
+	//  "_tmp": {
+	//    "type": "object"
+	//  },
+	if len(definition) != 1 {
+		return false
+	}
+	return mappingParameter("type", definition) == "object"
+}
+
+func isObject(definition mappingDefinitions) bool {
+	// Example:
+	// "http": {
+	//   "properties": {
+	// 	   "request": {
+	// 	     "properties": {
+	//         "method": {
+	//           "type": "keyword",
+	//           "ignore_above": 1024
+	//         }
+	//       }
+	//     }
+	//   }
+	// }
+	if len(definition) > 2 {
+		// there could also be a "dynamic: true" in the mapping
+		return false
+	}
+	field, ok := definition["properties"]
+	if !ok {
+		return false
+	}
+	if _, ok = field.(map[string]any); !ok {
+		return false
+	}
```

Comment:
Not sure why, but doing a type assertion to `mappingDefinitions` directly here (and in other places) did not work as I expected. I had to use `map[string]any`.

Reply:
Decided to remove the `mappingDefinitions` struct as suggested in https://github.com/elastic/elastic-package/pull/2214/files#r1847123096